Sync `rust-analyzer`, add `rust-analyzer-proc-macro-srv` binary to Rustc component

fasterthanlime opened the pull request:
As discussed earlier with @jyn514 and @pietroalbini, I'm also going to use this PR to have `dist::Rustc` build the `rust-analyzer-proc-macro-srv` binary introduced in:

  * https://github.com/rust-lang/rust-analyzer/pull/12871

rust-highfive commented:
r? @Mark-Simulacrum

(rust-highfive has picked a reviewer for you, use r? to override)

fasterthanlime commented:
r? @jyn514

jyn514 commented:
> I'm also going to use this PR to have dist::Rustc build the rust-analyzer-proc-macro-srv binary

I don't see that change here, I assume that's why it's still a draft.

@rustbot author

fasterthanlime commented:
Ok, so! I added `tool::RustAnalyzerProcMacroSrv`, and I ensure'd it from `dist::Rustc`, and I checked that it works by doing:

```shell
$ ./x.py dist rustc
(cut)
$ tar wtf build/dist/rustc-1.64.0-dev-x86_64-unknown-linux-gnu.tar.gz | sed -n 's/rustc-[^/]*\/rustc//p' | grep proc-macro-srv
/libexec/rust-analyzer-proc-macro-srv
```

Things that bother me are:

  * This isn't covered by `x.py check` right now
  * I'm not 100% confident about any of the bootstrap code I added - I don't get code intelligence in there via rust-analyzer, so it's mostly grepping and guessing

fasterthanlime commented:
@rustbot review

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/7522337380?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/f6b3960b956e40d74e773d6dcfc14764c468ac9b/checks/7522337380/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
---
extracting /checkout/obj/build/cache/2022-07-21/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/stage0
skip untracked path cpu-usage.csv during rustfmt invocations
Diff in /checkout/src/bootstrap/dist.rs at line 362:
 
             builder.install(&builder.rustdoc(compiler), &image.join("bin"), 0o755);
 
-            let ra_proc_macro_srv = builder.ensure(tool::RustAnalyzerProcMacroSrv {
-                target: compiler.host,
-                target: compiler.host,
-            }).expect("rust-analyzer-proc-macro-server always builds");
+            let ra_proc_macro_srv = builder
+                .ensure(tool::RustAnalyzerProcMacroSrv { compiler, target: compiler.host })
+                .expect("rust-analyzer-proc-macro-server always builds");
             builder.install(&ra_proc_macro_srv, &image.join("libexec"), 0o755);
             let libdir_relative = builder.libdir_relative(compiler);
             let libdir_relative = builder.libdir_relative(compiler);
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/src/bootstrap/native.rs" "/checkout/src/bootstrap/bin/sccache-plus-cl.rs" "/checkout/src/bootstrap/dist.rs" "/checkout/src/bootstrap/bin/llvm-config-wrapper.rs" "/checkout/src/bootstrap/config.rs" "/checkout/src/bootstrap/bin/rustc.rs" "/checkout/src/bootstrap/format.rs" "/checkout/library/proc_macro/src/bridge/mod.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
thread '<unnamed>' panicked at 'thread '<unnamed>' panicked at 'tx.send(entry.into_path()) failed with sending on a closed channel', format.rs:166:17
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
tx.send(entry.into_path()) failed with sending on a closed channel', format.rs:166:17
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Any { .. }', /cargo/registry/src/github.com-1ecc6299db9ec823/ignore-0.4.18/src/walk.rs:1302:31
```

</details>

bjorn3 commented:
> I don't get code intelligence in there via rust-analyzer, so it's mostly grepping and guessing

Try setting `rust-analyzer.linkedProjects` to `["./Cargo.toml", "src/bootstrap/Cargo.toml"]`.

jyn514 commented:
>No need for a separate check, it's covered by check::RustAnalyzer which checks the whole workspace

Does that mean we don't need a separate `tool` step either? since it's built as part of `tool::RustAnalyzer`?

fasterthanlime commented:
> since it's built as part of `tool::RustAnalyzer`?

Nope, that one builds `src/tools/rust-analyzer/crates/rust-analyzer`, not `src/tools/rust-analyzer`.

fasterthanlime commented:
I made the two bootstrap tests pass in https://github.com/rust-lang/rust/pull/99756/commits/e0750037f305452ea80066ae371341f29608d3c0 but I'm not sure if:

  * it's normal that the test "fixtures" needed to be updated since we're now building More Stuff
  * that means I did something wrong (picked the wrong target or something)

These lines I don't feel super confident about:

https://github.com/rust-lang/rust/blob/e0750037f305452ea80066ae371341f29608d3c0/src/bootstrap/dist.rs#L365-L368

Mark-Simulacrum commented on src/bootstrap/dist.rs:
```diff
@@ -362,6 +362,11 @@ impl Step for Rustc {
+            let ra_proc_macro_srv = builder
+                .ensure(tool::RustAnalyzerProcMacroSrv { compiler, target: compiler.host })
+                .expect("rust-analyzer-proc-macro-server always builds");
+            builder.install(&ra_proc_macro_srv, &image.join("libexec"), 0o755);
```
So I think the problem here is that this is using the `compiler` that this step uses/produces (`run.builder.compiler(run.builder.top_stage, run.target)`), which isn't quite right for compiling other things. (Since it means that we need to build a std using that compiler, which isn't otherwise typically necessary).

I think the fix here is going to be to change compiler in the tool invocation to be closer to what is used for e.g. Miri compilation elsewhere -- with slight modifications: `builder.compiler_for(compiler.stage, builder.config.build, compiler.host)` looks right to me. My hope is that with that change the tests will no longer need modification.

fasterthanlime replied:
That seems to have done it! https://github.com/rust-lang/rust/pull/99756/commits/4c4ea87729cbd964e4263499c41149d24c8cbb00 applies your suggested changes, and https://github.com/rust-lang/rust/pull/99756/commits/55b8ef4d9a4024e638aa71ffa1dfedcff41081fb reverses the test changes.

Mark-Simulacrum commented:
If possible, please squash the bootstrap-related commits into just one, and then I'm happy to r+

Thanks!

fasterthanlime commented:
> If possible, please squash the bootstrap-related commits into just one, and then I'm happy to r+

@Mark-Simulacrum All done!

I went as far as doing an `x.py dist rustc` from Windows to check that it produces the expected results, and it does:

```shell
❯ tar wtf .\build\dist\rustc-1.64.0-dev-x86_64-pc-windows-msvc.tar.gz | grep proc-macro-srv
rustc-1.64.0-dev-x86_64-pc-windows-msvc/rustc/libexec/rust-analyzer-proc-macro-srv.exe
```

Mark-Simulacrum commented:
@bors r+ rollup=iffy

bors commented:
:pushpin: Commit 6ea7d82dcc83d71753189b5f4ce676a31a0e88ab has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum 6ea7d82dcc83d71753189b5f4ce676a31a0e88ab -->
<!-- homu: {"type":"Approved","sha":"6ea7d82dcc83d71753189b5f4ce676a31a0e88ab","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 6ea7d82dcc83d71753189b5f4ce676a31a0e88ab with merge 05e678ccca974a8d0c26991083fb4cf8fff84e74...
<!-- homu: {"type":"BuildStarted","head_sha":"6ea7d82dcc83d71753189b5f4ce676a31a0e88ab","merge_sha":"05e678ccca974a8d0c26991083fb4cf8fff84e74"} -->

rust-timer commented:
Finished benchmarking commit (05e678ccca974a8d0c26991083fb4cf8fff84e74): [comparison url](https://perf.rust-lang.org/compare.html?start=48316dfea1914d25189fa441e7310449ed76a446&end=05e678ccca974a8d0c26991083fb4cf8fff84e74&stat=instructions:u).


### [Instruction count](https://perf.rust-lang.org/compare.html?start=48316dfea1914d25189fa441e7310449ed76a446&end=05e678ccca974a8d0c26991083fb4cf8fff84e74&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=48316dfea1914d25189fa441e7310449ed76a446&end=05e678ccca974a8d0c26991083fb4cf8fff84e74&stat=max-rss)
<details>
<summary>Results</summary>


- Primary benchmarks: 🎉 relevant improvement found
- Secondary benchmarks: 😿 relevant regression found

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions 😿 <br /> (primary) | N/A      | N/A | 0         |
| Regressions 😿 <br /> (secondary) | 2.5%     | 2.5% | 1         |
| Improvements 🎉 <br /> (primary) | -2.0%    | -2.0% | 1         |
| Improvements 🎉 <br /> (secondary) | N/A      | N/A | 0         |
| All 😿🎉 (primary) | -2.0%    | -2.0% | 1         |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=48316dfea1914d25189fa441e7310449ed76a446&end=05e678ccca974a8d0c26991083fb4cf8fff84e74&stat=cycles:u)
<details>
<summary>Results</summary>


- Primary benchmarks: mixed results
- Secondary benchmarks: no relevant changes found

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions 😿 <br /> (primary) | 2.2%     | 2.2% | 1         |
| Regressions 😿 <br /> (secondary) | N/A      | N/A | 0         |
| Improvements 🎉 <br /> (primary) | -2.5%    | -2.5% | 1         |
| Improvements 🎉 <br /> (secondary) | N/A      | N/A | 0         |
| All 😿🎉 (primary) | -0.2%    | -2.5% | 2         |
</details>

[^1]: *the arithmetic mean of the percent change*
[^2]: *number of relevant changes*

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).



@rustbot label: -perf-regression
